packages/react-aria-components/src/Tree.tsx

```diff
@@ -273,7 +291,87 @@ function TreeInner<T extends object>({props, collection, treeRef: ref}: TreeInne
           getDropTargetFromPoint(x, y, isValidDropTarget) {
             let target = dropTargetDelegate.getDropTargetFromPoint(x, y, isValidDropTarget);
+            let tracking = pointerTrackingRef.current;
+            
+            // Calculate movement direction
+            let deltaY = y - tracking.lastY;
+            let currentMovement: 'up' | 'down' | null = null;
+            
+            if (Math.abs(deltaY) > 3) {
+              currentMovement = deltaY > 0 ? 'down' : 'up';
+              tracking.movementDirection = currentMovement;
+            }
+            
+            // Handle ambiguous drop position: 'after last child' or 'after parent'
             if (target?.type === 'item' && target.dropPosition === 'after') {
+              let item = state.collection.getItem(target.key);
+              let parentKey = item?.parentKey;
+              
+              if (parentKey) {
+                let parentItem = state.collection.getItem(parentKey);
+                let isParentExpanded = parentItem && state.expandedKeys.has(parentKey);
+                
+                if (isParentExpanded) {
+                  let nextKey = state.collection.getKeyAfter(target.key);
+                  let nextItem = nextKey ? state.collection.getItem(nextKey) : null;
+                  let isLastChild = !nextItem || nextItem.parentKey !== parentKey;
+                  
+                  if (isLastChild) {                    
+                    let afterParentTarget = {
+                      type: 'item',
+                      key: parentKey,
+                      dropPosition: 'after'
+                    } as const;
+                    
+                    // eslint-disable-next-line max-depth
+                    if (!tracking.boundaryContext || tracking.boundaryContext.parentKey !== parentKey) {
+                      let initialPreference: 'inside' | 'after' = 'inside';
+                      // eslint-disable-next-line max-depth
+                      if (tracking.movementDirection === 'up') {
+                        initialPreference = 'after';
+                      } else if (tracking.movementDirection === 'down') {
+                        initialPreference = 'inside';
+                      }
+                      
+                      tracking.boundaryContext = {
+                        parentKey,
+                        lastChildKey: target.key,
+                        preferredPosition: initialPreference,
+                        lastSwitchY: y,
+                        entryDirection: tracking.movementDirection
+                      };
+                    }
+                    
+                    let context = tracking.boundaryContext;
+                    let distanceFromLastSwitch = Math.abs(y - context.lastSwitchY);
+                    
+                    // Switch logic based on continued movement direction
+                    if (distanceFromLastSwitch > 12) { // Threshold for smooth switching
```

Comment:
Hard-coded now, but maybe this should be relative to the item height? Or be dynamic based on input type (touch or mouse)